src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/Source.cs

```diff
@@ -0,0 +1,238 @@
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Roslyn.Utilities;
+using AsyncCompletionData = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using RoslynTrigger = Microsoft.CodeAnalysis.Completion.CompletionTrigger;
+using VSCompletionItem = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.AsyncCompletion
+{
+    internal class Source : IAsyncCompletionSource
+    {
+        internal const string RoslynItem = nameof(RoslynItem);
+        internal const string TriggerSnapshot = nameof(TriggerSnapshot);
+        internal const string InsertionText = nameof(InsertionText);
+        internal const string HasSuggestionItemOptions = nameof(HasSuggestionItemOptions);
+        internal const string Description = nameof(Description);
+
+        private static readonly ImmutableArray<ImageElement> s_WarningImageAttributeImagesArray = 
+            ImmutableArray.Create(new ImageElement(Glyph.CompletionWarning.GetImageId(), EditorFeaturesResources.Warning_image_element_automation_name));
+
+        public AsyncCompletionData.CompletionStartData InitializeCompletion(AsyncCompletionData.CompletionTrigger trigger, SnapshotPoint triggerLocation, CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return AsyncCompletionData.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return AsyncCompletionData.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var sourceText = document.GetTextSynchronously(cancellationToken);
+
+            // TODO: Check CompletionOptions.TriggerOnTyping  https://github.com/dotnet/roslyn/issues/27427
+            if (trigger.Reason != AsyncCompletionData.CompletionTriggerReason.Invoke &&
+                trigger.Reason != AsyncCompletionData.CompletionTriggerReason.InvokeAndCommitIfUnique && 
```

Comment:
why do we need to check the reasons here?  we already pass the trigger information into service.ShouldTriggerCompletion.  it is already checking that stuff.